c10/core/Event.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include "ATen/Context.h"
+#include "c10/util/Optional.h"
+#include "c10/util/Exception.h"
+#include "c10/core/Stream.h"
+
+#include <memory>
+#include <limits>
+#include <atomic>
+#include <utility>
+
+namespace at { namespace cuda {  namespace detail {
+    struct CUDAHooks;
+}}} // at::cuda::detail
+
+namespace c10 {
+
+/*
+* A generic Event capable of representing device-specific events when the appropriate
+* backend is loaded. It is designed to dispatch calls to different backends
+* based on the Stream the event is recorded on.
+*
+* This class is not thread-safe and is movable but not copyable. 
+*
+*/
+struct Event final {
+
+static constexpr unsigned int DEFAULT_FLAGS = std::numeric_limits<unsigned int>::max();
+friend class at::cuda::detail::CUDAHooks;
```

Comment:
Ick.

Reply:
I'll get rid of this. See comment below (two birds one stone).